### PR TITLE
[WebAuthn] Add tests for cross-origin create()

### PR DESCRIPTION
**Pull request by stephenmcgruer**
See https://github.com/w3c/webauthn/pull/1801

### CODE REVIEW

**Comment by stephenmcgruer** [comment]
This test should pass in WPT's Chrome Dev setup once it reaches 122.0.6196.0 (which may not be until after the holidays). It passes locally for me using Chrome Canary. I'm happy to wait until Chrome Dev catches up before landing, if desirable :)

File: webauthn/createcredential-cross-origin-iframe.https.sub.html [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        });
+    }
+
+    function createIframe(test) {
+        const iframeElement = document.createElement("iframe");
+        document.body.appendChild(iframeElement);
+        test.add_cleanup(() => {
+            iframeElement.remove();
+        });
+        return iframeElement;
```
Comment (nsatragno) [inline webauthn/createcredential-cross-origin-iframe.https.sub.html:39]: Consider making `createIframe` take `src` and `allow`, and return a promise that resolves to the `iframe` once `subframe-loaded` is received. That way you don't have add code to wait for the iframe to load on every test.

File: webauthn/createcredential-cross-origin-iframe.https.sub.html [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        await loadedPromise;
+
+        const resultPromise = getMessageData("result");
+        iframe.contentWindow.postMessage({"type": "create-credential", "addUserActivation": false}, {targetOrigin: targetOrigin});
+        const data = await resultPromise;
+
+        assert_equals(data.result, "failure");
+    }, "create() in cross-origin iframe fails with permissions policy but no user activation");
+
+    promise_test(async (test) => {
```
Comment (nsatragno) [inline webauthn/createcredential-cross-origin-iframe.https.sub.html:69]: Should we have a test that verifies that a gesture on the parent frame cannot trigger webauthn on the subframe?

File: webauthn/createcredential-cross-origin-iframe.https.sub.html [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    promise_test(async (test) => {
+        const iframe = createIframe(test);
+        const loadedPromise = getMessageData("subframe-loaded");
+        iframe.src = `${targetOrigin}/webauthn/resources/webauthn-subframe.sub.html`;
+        await loadedPromise;
+
+        const resultPromise = getMessageData("result");
+        iframe.contentWindow.postMessage({"type": "create-credential"}, {targetOrigin: targetOrigin});
```
Comment (nsatragno) [inline webauthn/createcredential-cross-origin-iframe.https.sub.html:49]: Pedantic nit: let's choose whether we wrap object keys between quotes or not, and stick to that. My vote is no quotes.

File: webauthn/createcredential-cross-origin-iframe.https.sub.html [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        const data = await resultPromise;
+
+        assert_equals(data.result, "failure");
```
Comment (nsatragno) [inline webauthn/createcredential-cross-origin-iframe.https.sub.html:52]: Let's validate that the error returned is `NotAllowedError`.

File: webauthn/resources/webauthn-subframe.sub.html [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            reportResult("success");
+        } catch (e) {
+            reportResult("failure", e.message);
+        }
+    }
+
+    window.addEventListener("message", e => {
+        if (e.data.type == "create-credential") {
+            createCredentialAndReportResult(e.data.addUserActivation);
+        }
```
Comment (nsatragno) [inline webauthn/resources/webauthn-subframe.sub.html:31]: We should throw here if we get an unrecognized message.

**Review by nsatragno — COMMENTED** [review]
Thanks for the WPT!

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 1 above)
Comment (stephenmcgruer) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html:51]: Done, thanks!

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 3 above)
Comment (stephenmcgruer) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html]: Haha, whoops, sorry. Resolved to no quotes.

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 4 above)
Comment (stephenmcgruer) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html:62]: Done.

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 2 above)
Comment (stephenmcgruer) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html:77]: Done

File: webauthn/resources/webauthn-subframe.sub.html (hunk 5 above)
Comment (stephenmcgruer) [reply webauthn/resources/webauthn-subframe.sub.html:32]: Done, though it won't be directly visible in tests sadly - one has to check the console and/or devtools logs.

**Comment by stephenmcgruer** [comment]
Results from wpt.fyi - https://wpt.fyi/results/webauthn/createcredential-cross-origin-iframe.https.sub.html?diff&filter=ADC&run_id=5098614518448128&run_id=5186312449032192

**Comment by stephenmcgruer** [comment]
@kenrb if you can take a look in Nina's absence, that would be great thanks!

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 4 above)
Comment (kenrb) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html:62]: This doesn't appear to have been done in this test, although it was added for the other ones?

**Review by kenrb — COMMENTED** [review]
lg although one change might have been missed?

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 4 above)
Comment (stephenmcgruer) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html:62]: Whoops, good catch. Fixed!

**Comment by stephenmcgruer** [comment]
Thanks Ken! Assigning to @rsolomakhin for LGTM, since looks like Ken maybe isn't a reviewing member of WPT?

**Review by rsolomakhin — APPROVED** [review]

**Comment by stephenmcgruer** [comment]
The failure from Firefox stability is from the mozilla side:

```
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://hg.mozilla.org/mozilla-central/archive/tip.zip/testing/profiles/
```

So I'm going to merge this.

**Review by sarobakhtiyari — APPROVED** [review]

**Review by sarobakhtiyari — APPROVED** [review]

**Review by sarobakhtiyari — APPROVED** [review]

**Review by sarobakhtiyari — APPROVED** [review]

**Review by sarobakhtiyari — APPROVED** [review]

File: webauthn/createcredential-cross-origin-iframe.https.sub.html (hunk 4 above)
Comment (sarobakhtiyari) [reply webauthn/createcredential-cross-origin-iframe.https.sub.html:62]: - [ ]

File: webauthn/resources/webauthn-subframe.sub.html (hunk 5 above)
Comment (sarobakhtiyari) [reply webauthn/resources/webauthn-subframe.sub.html:32]: - [ ] > - - [ ] # - [ ] @~~@~~@ ~~ ~~

**Pull request merged**